leases/tasks/send_invoices.py

```diff
@@ -0,0 +1,231 @@
+        end_date__year=lease_year,
+    )
+
+    return leases
+
+
+def create_new_lease(
+    lease: Union[BerthLease, WinterStorageLease], start_date: date, end_date: date
+) -> Union[BerthLease, WinterStorageLease]:
+    """Create a new lease instance"""
+
+    # Blank the PK to signal that a new instance has to be created
+    lease.pk = None
+
+    # Manually need to set creating because of the checks performed on clean
+    lease._state.adding = True
+
+    # Update the dates for the lease for the next season
+    lease.start_date = start_date
+    lease.end_date = end_date
+    lease.save()
+
+    return lease
+
+
+def send_berth_invoices(request: HttpRequest, due_date: date = None) -> Dict[str, List]:
+    return send_invoices(
+        request,
+        BERTH_LEASES,
+        season_start=calculate_season_start_date(),
+        season_end=calculate_season_end_date(),
+        due_date=due_date,
+    )
+
+
+def send_winter_storage_invoices(
+    request: HttpRequest, due_date: date = None
+) -> Dict[str, List]:
+    raise NotImplementedError
+
+
+@transaction.atomic
+def send_invoices(
+    request: HttpRequest,
+    lease_type: str,
+    season_start: date,
+    season_end: date,
+    due_date: date = None,
+) -> Dict[str, List]:  # noqa: C901
+    successful_orders: List[UUID] = []
+    failed_orders: List[Dict[UUID, str]] = []
+    failed_leases: List[Dict[UUID, str]] = []
+
+    # Default the due date to 14 days from the date when the task is executed
+    if not due_date:
+        due_date = today() + relativedelta(days=14)
+
+    leases = get_valid_leases(season_start, lease_type)
+
+    orders: List[Order] = []
+
+    for lease in leases:
+        lease = create_new_lease(lease, season_start, season_end)
+
+        # Get the associated product
+        product = get_product(lease)
+
+        # If no product is found, the billing can't be done
+        if not product:
+            fail_lease(lease, _("No suitable product found"), failed_leases)
+            continue
+
+        try:
+            order = Order.objects.create(
+                customer=lease.customer, lease=lease, product=product
+            )
+            orders.append(order)
+        except ValidationError as e:
+            fail_lease(lease, str(e), failed_leases)
+
+    # Fetch all the profiles from the Profile service
+    profiles: Dict[UUID, HelsinkiProfileUser] = ProfileService(
+        request
+    ).get_all_profiles()
+
+    # Go through all the created orders
+    for order in orders:
+        # If the profile doesn't have an associated email or if it has an example.com address, set it as failed
+        email = profiles.get(order.customer.id).email
+        if email and "example.com" not in email:
+            try:
+                approve_order(order, email, due_date, request)
+                successful_orders.append(order.id)
+            except (
+                AnymailError,
+                OSError,
+                ValidationError,
```

Comment:
Are ValidationError, VenepaikkaGraphQLError and OSError here "expected" as part of situation-as-usual, or are they more like an indication that something is horribly wrong and someone should take a look before the process can continue?
Maybe log the stacktrace to ease debugging?

Reply:
`VenepaikkaGraphQLError` would be a more "situation-as-usual" error handling (something went wrong with the instance.

`ValidationError` would be a bit more concerning, since it's raised by the `get_order_notification_type` meaning that an order doesn't have a valid type, but shouldn't be the case very often.

IIRC, `OSError` can be raised by `anymail`/`django-mailer`/`django-ilmoitin`, so probably not critical (systems down-critical), but also worth checking.

In any case, definitely worth logging this more "serious" errors. 👍